File: src/kicad_sch.py
"""
KiCad schematic (.kicad_sch) parser.

Extracts components, wires, junctions, and labels from KiCad schematics.
"""

from dataclasses import dataclass, field
from typing import Optional
import math

from sexpr import parse, get, get_all, get_value, SExpr


@dataclass
class Point:
    x: float
    y: float

    def __hash__(self):
        return hash((round(self.x, 4), round(self.y, 4)))

    def __eq__(self, other):
        if not isinstance(other, Point):
            return False
        return (round(self.x, 4) == round(other.x, 4) and
                round(self.y, 4) == round(other.y, 4))


@dataclass
class Pin:
    number: str
    name: str
    x: float  # Relative to symbol origin
    y: float
    angle: float  # Pin angle in degrees (0=right, 90=up, 180=left, 270=down)
    pin_type: str  # input, output, bidirectional, passive, power_in, etc.


@dataclass
class LibSymbol:
    """Symbol definition from lib_symbols section."""
    lib_id: str
    pins: list[Pin] = field(default_factory=list)
    is_power: bool = False


@dataclass
class Component:
    """A placed component instance."""
    lib_id: str
    reference: str
    value: str
    footprint: str
    x: float
    y: float
    angle: float  # Rotation in degrees
    mirror: Optional[str]  # None, 'x', or 'y'
    unit: int
    dnp: bool
    uuid: str
    pins: dict[str, Point] = field(default_factory=dict)  # pin_number -> absolute position


@dataclass
class Wire:
    """A wire segment."""
    points: list[Point]
    uuid: str


@dataclass
class Junction:
    """A junction point where wires connect."""
    x: float
    y: float
    uuid: str


@dataclass
class Label:
    """A net label (local, global, or hierarchical)."""
    name: str
    x: float
    y: float
    angle: float
    label_type: str  # 'local', 'global', 'hierarchical'
    uuid: str


@dataclass
class Schematic:
    """Parsed KiCad schematic."""
    title: str
    lib_symbols: dict[str, LibSymbol] = field(default_factory=dict)
    components: list[Component] = field(default_factory=list)
    wires: list[Wire] = field(default_factory=list)
    junctions: list[Junction] = field(default_factory=list)
    labels: list[Label] = field(default_factory=list)


def parse_schematic(filepath: str) -> Schematic:
    """Parse a KiCad schematic file."""
    with open(filepath, 'r') as f:
        text = f.read()

    expr = parse(text)
    return parse_schematic_expr(expr)


def parse_schematic_expr(expr: SExpr) -> Schematic:
    """Parse schematic from S-expression."""
    if not isinstance(expr, list) or expr[0] != 'kicad_sch':
        raise ValueError("Not a valid KiCad schematic")

    sch = Schematic(title='')

    # Parse title block
    title_block = get(expr, 'title_block')
    if title_block:
        sch.title = get_value(title_block, 'title') or ''

    # Parse library symbols
    lib_symbols = get(expr, 'lib_symbols')
    if lib_symbols:
        for sym_expr in get_all(lib_symbols, 'symbol'):
            lib_sym = parse_lib_symbol(sym_expr)
            if lib_sym:
                sch.lib_symbols[lib_sym.lib_id] = lib_sym

    # Parse component instances
    for sym_expr in get_all(expr, 'symbol'):
        comp = parse_component(sym_expr, sch.lib_symbols)
        if comp:
            sch.components.append(comp)

    # Parse wires
    for wire_expr in get_all(expr, 'wire'):
        wire = parse_wire(wire_expr)
        if wire:
            sch.wires.append(wire)

    # Parse junctions
    for junc_expr in get_all(expr, 'junction'):
        junc = parse_junction(junc_expr)
        if junc:
            sch.junctions.append(junc)

    # Parse labels
    for label_expr in get_all(expr, 'label'):
        label = parse_label(label_expr, 'local')
        if label:
            sch.labels.append(label)

    for label_expr in get_all(expr, 'global_label'):
        label = parse_label(label_expr, 'global')
        if label:
            sch.labels.append(label)

    for label_expr in get_all(expr, 'hierarchical_label'):
        label = parse_label(label_expr, 'hierarchical')
        if label:
            sch.labels.append(label)

    return sch


def parse_lib_symbol(expr: SExpr) -> Optional[LibSymbol]:
    """Parse a library symbol definition."""
    if not isinstance(expr, list) or len(expr) < 2:
        return None

    lib_id = expr[1]
    if not isinstance(lib_id, str):
        return None

    # Check if it's a power symbol
    is_power = get(expr, 'power') is not None

    # Find all pins - they're in subsymbols like "SymName_1_1"
    pins = []
    for sub_sym in get_all(expr, 'symbol'):
        if isinstance(sub_sym, list) and len(sub_sym) > 1:
            # Parse pins within subsymbol
            for pin_expr in get_all(sub_sym, 'pin'):
                pin = parse_pin(pin_expr)
                if pin:
                    pins.append(pin)

    return LibSymbol(lib_id=lib_id, pins=pins, is_power=is_power)


def parse_pin(expr: SExpr) -> Optional[Pin]:
    """Parse a pin definition from a library symbol."""
    if not isinstance(expr, list) or len(expr) < 3:
        return None

    # Format: (pin TYPE STYLE (at X Y ANGLE) (length L) (name "NAME" ...) (number "NUM" ...))
    pin_type = expr[1] if len(expr) > 1 else 'passive'

    at = get(expr, 'at')
    if not at or len(at) < 4:
        return None

    x = float(at[1])
    y = float(at[2])
    angle = float(at[3]) if len(at) > 3 else 0

    name_expr = get(expr, 'name')
    name = name_expr[1] if name_expr and len(name_expr) > 1 else ''

    number_expr = get(expr, 'number')
    number = number_expr[1] if number_expr and len(number_expr) > 1 else ''

    return Pin(number=str(number), name=str(name), x=x, y=y, angle=angle, pin_type=str(pin_type))


def parse_component(expr: SExpr, lib_symbols: dict[str, LibSymbol]) -> Optional[Component]:
    """Parse a component instance."""
    if not isinstance(expr, list):
        return None

    lib_id_expr = get(expr, 'lib_id')
    if not lib_id_expr or len(lib_id_expr) < 2:
        return None
    lib_id = lib_id_expr[1]

    at = get(expr, 'at')
    if not at or len(at) < 3:
        return None
    x = float(at[1])
    y = float(at[2])
    angle = float(at[3]) if len(at) > 3 else 0

    # Check for mirror
    mirror = None
    mirror_expr = get(expr, 'mirror')
    if mirror_expr and len(mirror_expr) > 1:
        mirror = mirror_expr[1]

    unit = 1
    unit_expr = get(expr, 'unit')
    if unit_expr and len(unit_expr) > 1:
        unit = int(unit_expr[1])

    dnp = False
    dnp_expr = get(expr, 'dnp')
    if dnp_expr and len(dnp_expr) > 1:
        dnp = dnp_expr[1] == 'yes'

    uuid = ''
    uuid_expr = get(expr, 'uuid')
    if uuid_expr and len(uuid_expr) > 1:
        uuid = uuid_expr[1]

    # Extract properties
    reference = ''
    value = ''
    footprint = ''

    for prop in get_all(expr, 'property'):
        if len(prop) >= 3:
            prop_name = prop[1]
            prop_value = prop[2]
            if prop_name == 'Reference':
                reference = prop_value
            elif prop_name == 'Value':
                value = prop_value
            elif prop_name == 'Footprint':
                footprint = prop_value

    comp = Component(
        lib_id=lib_id,
        reference=reference,
        value=value,
        footprint=footprint,
        x=x, y=y,
        angle=angle,
        mirror=mirror,
        unit=unit,
        dnp=dnp,
        uuid=uuid
    )

    # Calculate absolute pin positions
    # Try lib_name first (for renamed symbols), then fall back to lib_id
    lib_name_expr = get(expr, 'lib_name')
    lib_name = lib_name_expr[1] if lib_name_expr and len(lib_name_expr) > 1 else None

    lookup_key = lib_name if lib_name and lib_name in lib_symbols else lib_id
    if lookup_key in lib_symbols:
        lib_sym = lib_symbols[lookup_key]
        for pin in lib_sym.pins:
            abs_pos = transform_pin(pin.x, pin.y, x, y, angle, mirror)
            comp.pins[pin.number] = abs_pos

    return comp


def transform_pin(pin_x: float, pin_y: float,
                  comp_x: float, comp_y: float,
                  angle: float, mirror: Optional[str]) -> Point:
    """Transform a pin position from symbol-relative to absolute coordinates.

    KiCad coordinate systems:
    - Schematic: Y increases downward
    - Symbol library: Y increases upward (standard math convention)

    So we need to flip Y when going from symbol to schematic coordinates.
    """
    # Flip Y axis (symbol coords to schematic coords)
    px, py = pin_x, -pin_y

    # Apply mirror
    if mirror == 'x':
        py = -py
    elif mirror == 'y':
        px = -px

    # Apply rotation (in schematic coordinate system)
    rad = math.radians(-angle)  # Negative because Y is flipped
    cos_a = math.cos(rad)
    sin_a = math.sin(rad)

    rx = px * cos_a - py * sin_a
    ry = px * sin_a + py * cos_a

    # Translate to component position
    return Point(comp_x + rx, comp_y + ry)


def parse_wire(expr: SExpr) -> Optional[Wire]:
    """Parse a wire segment."""
    if not isinstance(expr, list):
        return None

    pts = get(expr, 'pts')
    if not pts:
        return None

    points = []
    for xy in get_all(pts, 'xy'):
        if len(xy) >= 3:
            points.append(Point(float(xy[1]), float(xy[2])))

    uuid = ''
    uuid_expr = get(expr, 'uuid')
    if uuid_expr and len(uuid_expr) > 1:
        uuid = uuid_expr[1]

    return Wire(points=points, uuid=uuid) if points else None


def parse_junction(expr: SExpr) -> Optional[Junction]:
    """Parse a junction."""
    if not isinstance(expr, list):
        return None

    at = get(expr, 'at')
    if not at or len(at) < 3:
        return None

    uuid = ''
    uuid_expr = get(expr, 'uuid')
    if uuid_expr and len(uuid_expr) > 1:
        uuid = uuid_expr[1]

    return Junction(x=float(at[1]), y=float(at[2]), uuid=uuid)


def parse_label(expr: SExpr, label_type: str) -> Optional[Label]:
    """Parse a label (local, global, or hierarchical)."""
    if not isinstance(expr, list) or len(expr) < 2:
        return None

    name = expr[1]
    if not isinstance(name, str):
        return None

    at = get(expr, 'at')
    if not at or len(at) < 3:
        return None

    x = float(at[1])
    y = float(at[2])
    angle = float(at[3]) if len(at) > 3 else 0

    uuid = ''
    uuid_expr = get(expr, 'uuid')
    if uuid_expr and len(uuid_expr) > 1:
        uuid = uuid_expr[1]

    return Label(name=name, x=x, y=y, angle=angle, label_type=label_type, uuid=uuid)


if __name__ == '__main__':
    import sys
    if len(sys.argv) > 1:
        sch = parse_schematic(sys.argv[1])
        print(f"Title: {sch.title}")
        print(f"Library symbols: {len(sch.lib_symbols)}")
        print(f"Components: {len(sch.components)}")
        for c in sch.components:
            print(f"  {c.reference}: {c.lib_id} = {c.value} @ ({c.x}, {c.y}) rot={c.angle}")
            for pin_num, pos in c.pins.items():
                print(f"    Pin {pin_num}: ({pos.x:.2f}, {pos.y:.2f})")
        print(f"Wires: {len(sch.wires)}")
        print(f"Junctions: {len(sch.junctions)}")
        print(f"Labels: {len(sch.labels)}")
        for l in sch.labels:
            print(f"  {l.label_type}: {l.name} @ ({l.x}, {l.y})")

File: src/ts/kicadSch.ts
/**
 * KiCad schematic (.kicad_sch) parser.
 *
 * Extracts components, wires, junctions, and labels from KiCad schematics.
 */

import { parse, get, getAll, getValue, SExpr } from './sexpr';

export interface Point {
  x: number;
  y: number;
}

export function pointsEqual(a: Point, b: Point, tolerance: number = 0.0001): boolean {
  return (
    Math.abs(a.x - b.x) < tolerance &&
    Math.abs(a.y - b.y) < tolerance
  );
}

export function pointHash(p: Point): string {
  return `${p.x.toFixed(4)},${p.y.toFixed(4)}`;
}

export interface Pin {
  number: string;
  name: string;
  x: number; // Relative to symbol origin
  y: number;
  angle: number; // Pin angle in degrees (0=right, 90=up, 180=left, 270=down)
  pinType: string; // input, output, bidirectional, passive, power_in, etc.
}

export interface LibSymbol {
  libId: string;
  pins: Pin[];
  isPower: boolean;
}

export interface Component {
  libId: string;
  reference: string;
  value: string;
  footprint: string;
  x: number;
  y: number;
  angle: number; // Rotation in degrees
  mirror: 'x' | 'y' | null;
  unit: number;
  dnp: boolean;
  uuid: string;
  pins: Map<string, Point>; // pin_number -> absolute position
}

export interface Wire {
  points: Point[];
  uuid: string;
}

export interface Junction {
  x: number;
  y: number;
  uuid: string;
}

export interface Label {
  name: string;
  x: number;
  y: number;
  angle: number;
  labelType: 'local' | 'global' | 'hierarchical';
  uuid: string;
}

export interface Schematic {
  title: string;
  libSymbols: Map<string, LibSymbol>;
  components: Component[];
  wires: Wire[];
  junctions: Junction[];
  labels: Label[];
}

/**
 * Parse a KiCad schematic from text content.
 */
export function parseSchematic(text: string): Schematic {
  const expr = parse(text);
  return parseSchematicExpr(expr);
}

/**
 * Parse schematic from S-expression.
 */
function parseSchematicExpr(expr: SExpr): Schematic {
  if (!Array.isArray(expr) || expr[0] !== 'kicad_sch') {
    throw new Error('Not a valid KiCad schematic');
  }

  const sch: Schematic = {
    title: '',
    libSymbols: new Map(),
    components: [],
    wires: [],
    junctions: [],
    labels: [],
  };

  // Parse title block
  const titleBlock = get(expr, 'title_block');
  if (titleBlock) {
    sch.title = getValue(titleBlock, 'title') || '';
  }

  // Parse library symbols
  const libSymbols = get(expr, 'lib_symbols');
  if (libSymbols) {
    for (const symExpr of getAll(libSymbols, 'symbol')) {
      const libSym = parseLibSymbol(symExpr);
      if (libSym) {
        sch.libSymbols.set(libSym.libId, libSym);
      }
    }
  }

  // Parse component instances
  for (const symExpr of getAll(expr, 'symbol')) {
    const comp = parseComponent(symExpr, sch.libSymbols);
    if (comp) {
      sch.components.push(comp);
    }
  }

  // Parse wires
  for (const wireExpr of getAll(expr, 'wire')) {
    const wire = parseWire(wireExpr);
    if (wire) {
      sch.wires.push(wire);
    }
  }

  // Parse junctions
  for (const juncExpr of getAll(expr, 'junction')) {
    const junc = parseJunction(juncExpr);
    if (junc) {
      sch.junctions.push(junc);
    }
  }

  // Parse labels
  for (const labelExpr of getAll(expr, 'label')) {
    const label = parseLabel(labelExpr, 'local');
    if (label) {
      sch.labels.push(label);
    }
  }

  for (const labelExpr of getAll(expr, 'global_label')) {
    const label = parseLabel(labelExpr, 'global');
    if (label) {
      sch.labels.push(label);
    }
  }

  for (const labelExpr of getAll(expr, 'hierarchical_label')) {
    const label = parseLabel(labelExpr, 'hierarchical');
    if (label) {
      sch.labels.push(label);
    }
  }

  return sch;
}

/**
 * Parse a library symbol definition.
 */
function parseLibSymbol(expr: SExpr[]): LibSymbol | null {
  if (!Array.isArray(expr) || expr.length < 2) {
    return null;
  }

  const libId = expr[1];
  if (typeof libId !== 'string') {
    return null;
  }

  // Check if it's a power symbol
  const isPower = get(expr, 'power') !== undefined;

  // Find all pins - they're in subsymbols like "SymName_1_1"
  const pins: Pin[] = [];
  for (const subSym of getAll(expr, 'symbol')) {
    if (Array.isArray(subSym) && subSym.length > 1) {
      // Parse pins within subsymbol
      for (const pinExpr of getAll(subSym, 'pin')) {
        const pin = parsePin(pinExpr);
        if (pin) {
          pins.push(pin);
        }
      }
    }
  }

  return { libId, pins, isPower };
}

/**
 * Parse a pin definition from a library symbol.
 */
function parsePin(expr: SExpr[]): Pin | null {
  if (!Array.isArray(expr) || expr.length < 3) {
    return null;
  }

  // Format: (pin TYPE STYLE (at X Y ANGLE) (length L) (name "NAME" ...) (number "NUM" ...))
  const pinType = typeof expr[1] === 'string' ? expr[1] : 'passive';

  const at = get(expr, 'at');
  if (!at || at.length < 4) {
    return null;
  }

  const x = parseFloat(at[1] as string);
  const y = parseFloat(at[2] as string);
  const angle = at.length > 3 ? parseFloat(at[3] as string) : 0;

  const nameExpr = get(expr, 'name');
  const name = nameExpr && nameExpr.length > 1 && typeof nameExpr[1] === 'string'
    ? nameExpr[1]
    : '';

  const numberExpr = get(expr, 'number');
  const number = numberExpr && numberExpr.length > 1
    ? String(numberExpr[1])
    : '';

  return { number, name, x, y, angle, pinType };
}

/**
 * Parse a component instance.
 */
function parseComponent(expr: SExpr[], libSymbols: Map<string, LibSymbol>): Component | null {
  if (!Array.isArray(expr)) {
    return null;
  }

  const libIdExpr = get(expr, 'lib_id');
  if (!libIdExpr || libIdExpr.length < 2 || typeof libIdExpr[1] !== 'string') {
    return null;
  }
  const libId = libIdExpr[1];

  const at = get(expr, 'at');
  if (!at || at.length < 3) {
    return null;
  }
  const x = parseFloat(at[1] as string);
  const y = parseFloat(at[2] as string);
  const angle = at.length > 3 ? parseFloat(at[3] as string) : 0;

  // Check for mirror
  let mirror: 'x' | 'y' | null = null;
  const mirrorExpr = get(expr, 'mirror');
  if (mirrorExpr && mirrorExpr.length > 1) {
    const m = mirrorExpr[1];
    if (m === 'x' || m === 'y') {
      mirror = m;
    }
  }

  let unit = 1;
  const unitExpr = get(expr, 'unit');
  if (unitExpr && unitExpr.length > 1) {
    unit = parseInt(unitExpr[1] as string, 10);
  }

  let dnp = false;
  const dnpExpr = get(expr, 'dnp');
  if (dnpExpr && dnpExpr.length > 1) {
    dnp = dnpExpr[1] === 'yes';
  }

  let uuid = '';
  const uuidExpr = get(expr, 'uuid');
  if (uuidExpr && uuidExpr.length > 1 && typeof uuidExpr[1] === 'string') {
    uuid = uuidExpr[1];
  }

  // Extract properties
  let reference = '';
  let value = '';
  let footprint = '';

  for (const prop of getAll(expr, 'property')) {
    if (prop.length >= 3 && typeof prop[1] === 'string' && typeof prop[2] === 'string') {
      const propName = prop[1];
      const propValue = prop[2];
      if (propName === 'Reference') {
        reference = propValue;
      } else if (propName === 'Value') {
        value = propValue;
      } else if (propName === 'Footprint') {
        footprint = propValue;
      }
    }
  }

  const comp: Component = {
    libId,
    reference,
    value,
    footprint,
    x,
    y,
    angle,
    mirror,
    unit,
    dnp,
    uuid,
    pins: new Map(),
  };

  // Calculate absolute pin positions
  // Try lib_name first (for renamed symbols), then fall back to libId
  const libNameExpr = get(expr, 'lib_name');
  const libName = libNameExpr && libNameExpr.length > 1 && typeof libNameExpr[1] === 'string'
    ? libNameExpr[1]
    : null;

  const lookupKey = libName && libSymbols.has(libName) ? libName : libId;
  const libSym = libSymbols.get(lookupKey);

  if (libSym) {
    for (const pin of libSym.pins) {
      const absPos = transformPin(pin.x, pin.y, x, y, angle, mirror);
      comp.pins.set(pin.number, absPos);
    }
  }

  return comp;
}

/**
 * Transform a pin position from symbol-relative to absolute coordinates.
 *
 * KiCad coordinate systems:
 * - Schematic: Y increases downward
 * - Symbol library: Y increases upward (standard math convention)
 *
 * So we need to flip Y when going from symbol to schematic coordinates.
 */
function transformPin(
  pinX: number,
  pinY: number,
  compX: number,
  compY: number,
  angle: number,
  mirror: 'x' | 'y' | null
): Point {
  // Flip Y axis (symbol coords to schematic coords)
  let px = pinX;
  let py = -pinY;

  // Apply mirror
  if (mirror === 'x') {
    py = -py;
  } else if (mirror === 'y') {
    px = -px;
  }

  // Apply rotation (in schematic coordinate system)
  const rad = (-angle * Math.PI) / 180; // Negative because Y is flipped
  const cosA = Math.cos(rad);
  const sinA = Math.sin(rad);

  const rx = px * cosA - py * sinA;
  const ry = px * sinA + py * cosA;

  // Translate to component position
  return { x: compX + rx, y: compY + ry };
}

/**
 * Parse a wire segment.
 */
function parseWire(expr: SExpr[]): Wire | null {
  if (!Array.isArray(expr)) {
    return null;
  }

  const pts = get(expr, 'pts');
  if (!pts) {
    return null;
  }

  const points: Point[] = [];
  for (const xy of getAll(pts, 'xy')) {
    if (xy.length >= 3) {
      points.push({
        x: parseFloat(xy[1] as string),
        y: parseFloat(xy[2] as string),
      });
    }
  }

  let uuid = '';
  const uuidExpr = get(expr, 'uuid');
  if (uuidExpr && uuidExpr.length > 1 && typeof uuidExpr[1] === 'string') {
    uuid = uuidExpr[1];
  }

  return points.length > 0 ? { points, uuid } : null;
}

/**
 * Parse a junction.
 */
function parseJunction(expr: SExpr[]): Junction | null {
  if (!Array.isArray(expr)) {
    return null;
  }

  const at = get(expr, 'at');
  if (!at || at.length < 3) {
    return null;
  }

  let uuid = '';
  const uuidExpr = get(expr, 'uuid');
  if (uuidExpr && uuidExpr.length > 1 && typeof uuidExpr[1] === 'string') {
    uuid = uuidExpr[1];
  }

  return {
    x: parseFloat(at[1] as string),
    y: parseFloat(at[2] as string),
    uuid,
  };
}

/**
 * Parse a label (local, global, or hierarchical).
 */
function parseLabel(expr: SExpr[], labelType: 'local' | 'global' | 'hierarchical'): Label | null {
  if (!Array.isArray(expr) || expr.length < 2) {
    return null;
  }

  const name = expr[1];
  if (typeof name !== 'string') {
    return null;
  }

  const at = get(expr, 'at');
  if (!at || at.length < 3) {
    return null;
  }

  const x = parseFloat(at[1] as string);
  const y = parseFloat(at[2] as string);
  const angle = at.length > 3 ? parseFloat(at[3] as string) : 0;

  let uuid = '';
  const uuidExpr = get(expr, 'uuid');
  if (uuidExpr && uuidExpr.length > 1 && typeof uuidExpr[1] === 'string') {
    uuid = uuidExpr[1];
  }

  return { name, x, y, angle, labelType, uuid };
}
